Include/Rover/LinearRegressionModel.hpp

```diff
@@ -0,0 +1,80 @@
+#ifndef ROVER_LINEAR_REGRESSION_MODEL_HPP
+#define ROVER_LINEAR_REGRESSION_MODEL_HPP
+#include <type_traits>
+#include <vector>
+#include <dlib/matrix.h>
+
+namespace Rover {
+
+  //! Models a trial using linear regression.
+  /*!
+    \tparam T The Trial to model.
+  */
+  template<typename T>
+  class LinearRegressionModel {
+    public:
+
+      //! The Trial to model.
+      using Trial = T;
+
+      //! The result of the trial.
+      using Result = typename Trial::Sample::Result;
+
+      //! Constructs a linear regression model for a trial.
+      /*!
+        \param trial The trial to model.
+      */
+      LinearRegressionModel(const Trial& trial);
+
+      template<typename... Arg>
+      Result operator ()(Arg&&... arg) const;
+
+    private:
+      static dlib::matrix<Result> compute_transformation_vector(
+        const Trial& trial);
+      dlib::matrix<Result> m_transformation;
+  };
+
+  template<typename T>
+  LinearRegressionModel<T>::LinearRegressionModel(const T& trial)
+    : m_transformation(compute_transformation_vector(trial)) {}
+
+  template<typename T>
+  template<typename... Arg>
+  typename LinearRegressionModel<T>::Result
+      LinearRegressionModel<T>::operator ()(Arg&&... arg) const {
+    auto x = dlib::matrix<Result, 1, std::tuple_size_v<
+      typename Trial::Sample::Parameters> + 1>();
+    auto coef = std::vector<Result>();
```

Comment:
I can actually avoid this temporary vector too I think, but I do it the same way in compute_transformation_vector (where it is executed for every sample). Not sure whether it worth the complexity of the code?

Reply:
As a matter of fact, shouldn't we allow for perfect forwarding in the constructor to save on copies if a trial is consumed?
In practice though, almost always there won't be a move constructor as it only works for arithmetic types modeling real numbers. But maybe other trials following the same concept can benefit from that.